GVFS/GVFS.Common/Tracing/JsonTracer.cs

```diff
@@ -4,13 +4,15 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Linq;
+using System.Threading;
 
 namespace GVFS.Common.Tracing
 {
     public class JsonTracer : ITracer
     {
         public const string NetworkErrorEventName = "NetworkError";
 
+        private readonly ReaderWriterLockSlim listenersLock = new ReaderWriterLockSlim();
         private readonly List<EventListener> listeners;

```

Comment:
Is it an option to use a thread-safe collection (e.g. `ConcurrentBag`) rather than introducing a lock?  

If not, what are the scenario(s) where a thread-safe collection would be insufficient?

Reply:
There's a problem in `Dispose` where we enumerate `listeners` (calling `Dispose` on them in turn) and then clear the collection.

```csharp
foreach (EventListener listener in this.listeners)
{
    listener.Dispose();
}

this.listeners.Clear();
```


The enumeration with `foreach` is [fine because `GetEnumerator` returns a snapshot](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1.getenumerator?redirectedfrom=MSDN&view=netstandard-2.0#remarks), but when we later call `Clear` it's possible that a new listener has been added that is not going to be `Dispose`d of.

I could replace this with a loop doing `TryTake` while the collection is not empty as below..

```csharp
EventListener listener;
while (this.listeners.TryTake(out listener))
{
    listener.Dispose();
}
```

..but of course the failure of `TryTake` could be for reasons other than 'nothing left to take'/'collection is empty'.

This also raises the question about what exactly the point of `private bool disposed` is for in this class; we're not actually doing anything with it. Really we should be throwing `ObjectDisposedException` when attempting to interact with the `JsonTracer` (such as writing messages, or adding new listeners). To do that we'd also want to make the `this.disposed = true;` atomic with the clearing of the bag – we could do that by moving the `this.disposed = true` _before_ the `while(TryTake)`.

(In this case we'd also want to switch `private bool disposed` to `private int disposed` and use `Interlocked.CompareExchange` because there is no `bool` overload for CAS actions in .NET)

Reply:
> The enumeration with foreach is fine because GetEnumerator returns a snapshot, but when we later call Clear it's possible that a new listener has been added that is not going to be Disposed of.

I'm not sure that the lock is actually protecting us against missing a listener when we're disposing.

A call to add a new listener could execute immediately after the locked code block:

```
using (this.listenersLock.Read())
{
    foreach (EventListener listener in this.listeners)
    {
        listener.Dispose();
     }

    this.listeners.Clear();
}
```
And the new listener would not be disposed.

I have the same thoughts as @kewillford regarding performance, and unless we can show that we have a problem that can only be solved with locking I'd rather avoid the complexity introduced by adding locks.

I also think that with how specialized/critical our tracing code is, we'd be OK setting limitations on how callers are allowed to use `JsonTracer`.  For example, if either:

- `Dispose` is always called on the same thread as the thread adding listeners
-or-
- All listeners are added before sharing the `JsonTracer` with another thread

Then we might not need the complexity to handle the add/dispose race.

If we find we *do* need to handle that race though, then perhaps we could use a combination of locking in Add+Dispose and a thread-safe collection? 

That would allow lock-free reads of the collection and protection against crossing Adds and Dispose.

Reply:
I just saw that @kewillford is actually referring to a different lock 😄

Reply:
Sorry I left out an important part.  I do agree that the race condition you mentioned is best fixed with locking 👍

Reply:
> I'm not sure that the lock is actually protecting us against missing a listener when we're disposing.

This is exactly what I called out in my comment:

> The enumeration with foreach is fine because GetEnumerator returns a snapshot, **_but when we later call Clear it's possible that a new listener has been added that is not going to be Disposed of._**
>
> ...
>
> This also raises the question about what exactly the point of private bool disposed is for in this class; we're not actually doing anything with it. Really we should be throwing ObjectDisposedException when attempting to interact with the JsonTracer (such as writing messages, or adding new listeners). **_To do that we'd also want to make the this.disposed = true; atomic with the clearing of the bag – we could do that by moving the this.disposed = true before the while(TryTake)._**

The solution I call out first prevents new listeners from being added by setting the disposed flag (with the various `Add` methods also checking `!disposed` using `Interlocked.CompareExchange`), then it's safe to enumerate the listener collection disposing and removing them.

> I have the same thoughts as @kewillford regarding performance, and unless we can show that we have a problem that can only be solved with locking I'd rather avoid the complexity introduced by adding locks.
>
> I also think that with how specialized/critical our tracing code is, we'd be OK setting limitations on how callers are allowed to use JsonTracer. 

Let's just ignore `JsonTracer` being thread *un*safe for now if this is too difficult.

Reply:
> The enumeration with foreach is fine because GetEnumerator returns a snapshot, but when we later call Clear it's possible that a new listener has been added that is not going to be Disposed of.

Ahh sorry, I misread that point as something that is addressed with the current changes (that requires locking).

> This also raises the question about what exactly the point of private bool disposed is for in this class; we're not actually doing anything with it.

I saw that we are using `disposed` to detect if anyone is writing to a disposed tracer:

```
        private void WriteEvent(string eventName, EventLevel level, Keywords keywords, EventMetadata metadata, EventOpcode opcode)
        {
            string jsonPayload = metadata != null ? JsonConvert.SerializeObject(metadata) : null;

            if (this.disposed)
            {
                Console.WriteLine("Writing to disposed tracer");
                Console.WriteLine(jsonPayload);

                throw new ObjectDisposedException(nameof(JsonTracer));
            }
```

And I agree with the point you've raised here:

> In this case we'd also want to switch private bool disposed to private int disposed and use Interlocked.CompareExchange because there is no bool overload for CAS actions in .NET

`disposed` is not currently atomic and so threads might read the wrong value.

And I also agree that we're not consistently checking `disposed` in all the places we could.

> Let's just ignore `JsonTracer` being thread unsafe for now if this is too difficult

I think it's great that you called out that `listeners` is not thread safe, and I think it's worth either confirming that our usage of it is currently safe or fixing it as needed.

> this is still a draft PR for a reason

I missed that this is a draft PR ☹️ Apologies for jumping in with comments that were not at the right level  for a draft.

Reply:
A `ConcurrentBag` has been added in lieu of a lock here for `JsonTracer`. The `disposed` flag has been updated to now be read/written atomically.